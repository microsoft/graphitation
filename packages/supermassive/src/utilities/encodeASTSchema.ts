import {
  DirectiveDefinitionNode,
  EnumTypeDefinitionNode,
  FieldDefinitionNode,
  InputObjectTypeDefinitionNode,
  InputValueDefinitionNode,
  InterfaceTypeDefinitionNode,
  ObjectTypeDefinitionNode,
  ScalarTypeDefinitionNode,
  UnionTypeDefinitionNode,
  DocumentNode,
} from "graphql";
import {
  DirectiveDefinitionTuple,
  EnumTypeDefinitionTuple,
  FieldDefinitionTuple,
  InputObjectTypeDefinitionTuple,
  InputValueDefinitionRecord,
  InputValueDefinitionTuple,
  InterfaceTypeDefinitionTuple,
  ObjectTypeDefinitionTuple,
  ScalarTypeDefinitionTuple,
  UnionTypeDefinitionTuple,
  SchemaFragmentDefinitions,
  TypeKind,
  TypeReference,
} from "../schema/definition";
import { typeReferenceFromNode } from "../schema/reference";
import { valueFromASTUntyped } from "./valueFromASTUntyped";

export function encodeASTSchema(
  schemaFragment: DocumentNode,
): SchemaFragmentDefinitions {
  const types = Object.create(null);
  const directives = [];
  for (const definition of schemaFragment.definitions) {
    if (definition.kind === "ObjectTypeDefinition") {
      types[definition.name.value] = encodeObjectType(definition);
    } else if (definition.kind === "InputObjectTypeDefinition") {
      types[definition.name.value] = encodeInputObjectType(definition);
    } else if (definition.kind === "EnumTypeDefinition") {
      types[definition.name.value] = encodeEnumType(definition);
    } else if (definition.kind === "UnionTypeDefinition") {
      types[definition.name.value] = encodeUnionType(definition);
    } else if (definition.kind === "InterfaceTypeDefinition") {
      types[definition.name.value] = encodeInterfaceType(definition);
    } else if (definition.kind === "ScalarTypeDefinition") {
      types[definition.name.value] = encodeScalarType(definition);
    } else if (definition.kind === "DirectiveDefinition") {
      directives.push(encodeDirective(definition));
    }
  }
  return !directives.length ? { types } : { types, directives };
}

function encodeScalarType(
  _type: ScalarTypeDefinitionNode,
): ScalarTypeDefinitionTuple {
  return [TypeKind.SCALAR];
}

function encodeEnumType(node: EnumTypeDefinitionNode): EnumTypeDefinitionTuple {
  return [TypeKind.ENUM, (node.values ?? []).map((value) => value.name.value)];
}

function encodeObjectType(
  node: ObjectTypeDefinitionNode,
): ObjectTypeDefinitionTuple {
  const fields = Object.create(null);
  for (const field of node.fields ?? []) {
    fields[field.name.value] = encodeField(field);
  }
  if (!node.interfaces) {
    return [TypeKind.OBJECT, fields];
  }
  return [
    TypeKind.OBJECT,
    fields,
    node.interfaces.map((iface) => iface.name.value),
  ];
}

function encodeInterfaceType(
  node: InterfaceTypeDefinitionNode,
): InterfaceTypeDefinitionTuple {
  const fields = Object.create(null);
  for (const field of node.fields ?? []) {
    fields[field.name.value] = encodeField(field);
  }
  if (!node.interfaces) {
    return [TypeKind.INTERFACE, fields];
  }
  return [
    TypeKind.INTERFACE,
    fields,
    node.interfaces.map((iface) => iface.name.value),
  ];
}

function encodeUnionType(
  node: UnionTypeDefinitionNode,
): UnionTypeDefinitionTuple {
  return [TypeKind.UNION, (node.types ?? []).map((type) => type.name.value)];
}

function encodeInputObjectType(
  node: InputObjectTypeDefinitionNode,
): InputObjectTypeDefinitionTuple {
  const fields = Object.create(null);
  for (const field of node.fields ?? []) {
    fields[field.name.value] = encodeInputValue(field);
  }
  return [TypeKind.INPUT, fields];
}

function encodeField(
  node: FieldDefinitionNode,
): TypeReference | FieldDefinitionTuple {
  return !node.arguments?.length
    ? typeReferenceFromNode(node.type)
    : [typeReferenceFromNode(node.type), encodeArguments(node)];
}

function encodeArguments(
  node: FieldDefinitionNode | DirectiveDefinitionNode,
): InputValueDefinitionRecord {
  const args = Object.create(null);
  for (const argument of node.arguments ?? []) {
    args[argument.name.value] = encodeInputValue(argument);
  }
  return args;
}

function encodeInputValue(
  node: InputValueDefinitionNode,
): InputValueDefinitionTuple | TypeReference {
  if (!node.defaultValue) {
    return typeReferenceFromNode(node.type);
  }
  return [
    typeReferenceFromNode(node.type),
    valueFromASTUntyped(node.defaultValue),
  ];
}

function encodeDirective(
  node: DirectiveDefinitionNode,
): DirectiveDefinitionTuple {
  return !node.arguments?.length
    ? [node.name.value]
    : [node.name.value, encodeArguments(node)];
}
