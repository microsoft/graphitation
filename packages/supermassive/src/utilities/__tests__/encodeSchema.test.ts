import { encodeASTSchema } from "../encodeASTSchema";
import { parse } from "graphql";
import { readFileSync } from "fs";
import { join } from "path";

export const schemaSDL = parse(
  readFileSync(join(__dirname, "./fixtures/schema.graphql"), {
    encoding: "utf-8",
  }),
);

describe(encodeASTSchema, () => {
  test("correctly encodes sample AST schema", () => {
    const encoded = encodeASTSchema(schemaSDL);
    expect(encoded).toMatchSnapshot();
  });
});
