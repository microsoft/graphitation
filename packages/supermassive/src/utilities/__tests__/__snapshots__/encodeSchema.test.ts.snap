// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`encodeASTSchema correctly encodes sample AST schema 1`] = `
{
  "directives": [
    [
      "i18n",
      {
        "locale": 0,
      },
    ],
  ],
  "types": {
    "AdvancedInput": [
      6,
      {
        "enumField": "NodeType!",
        "otherField": 5,
      },
    ],
    "Alive": [
      4,
      [
        "Person",
        "Species",
      ],
    ],
    "Film": [
      2,
      {
        "characters": "[Person]",
        "created": 0,
        "director": 0,
        "edited": 0,
        "episode_id": 2,
        "id": 2,
        "opening_crawl": 0,
        "planets": "[Planet]",
        "producer": 0,
        "release_date": 0,
        "species": "[Species]",
        "starships": "[Starship]",
        "title": 5,
        "vehicles": "[Vehicle]",
      },
      [
        "Node",
      ],
    ],
    "Node": [
      3,
      {
        "id": 2,
      },
      [],
    ],
    "NodeType": [
      5,
      [
        "Person",
        "Starship",
        "Transport",
        "Species",
        "Vehicle",
        "Planet",
        "Film",
      ],
    ],
    "Person": [
      2,
      {
        "birth_year": 0,
        "created": 0,
        "edited": 0,
        "eye_color": 0,
        "films": "[Film]",
        "gender": 0,
        "hair_color": 0,
        "height": 2,
        "homeworld": "Planet",
        "id": 2,
        "image": 0,
        "mass": 2,
        "name": 0,
        "skin_color": 0,
        "starships": "[Starship]",
        "vehicles": "[Vehicle]",
      },
      [
        "Node",
      ],
    ],
    "Planet": [
      2,
      {
        "climate": 0,
        "created": 0,
        "diameter": 2,
        "edited": 0,
        "films": "[Film]",
        "gravity": 0,
        "id": 2,
        "name": 0,
        "orbital_period": 2,
        "population": 2,
        "residents": "[Person]",
        "rotation_period": 2,
        "surface_water": 0,
        "terrain": 0,
      },
      [
        "Node",
      ],
    ],
    "Query": [
      2,
      {
        "advancedDefaultInput": [
          0,
          {
            "input": [
              "AdvancedInput!",
              {
                "enumField": "Transport",
                "otherField": "Foo",
              },
            ],
          },
        ],
        "allFilms": "[Film]",
        "allPeople": "[Person]",
        "allPlanets": "[Planet]",
        "allSpecies": "[Species]",
        "allStarships": "[Starship]",
        "allTransports": "[Transport]",
        "film": [
          "Film",
          {
            "id": 7,
          },
        ],
        "multiArger": [
          0,
          {
            "a": 2,
            "b": 0,
            "c": "AdvancedInput",
          },
        ],
        "node": [
          "Node",
          {
            "id": 7,
            "nodeType": "NodeType!",
          },
        ],
        "person": [
          "Person",
          {
            "id": 7,
          },
        ],
        "planet": [
          "Planet",
          {
            "id": 7,
          },
        ],
        "search": [
          "[SearchResult]",
          {
            "search": 0,
          },
        ],
        "searchFilmsByTitle": [
          "[Film]",
          {
            "search": [
              5,
              "The Empire Strikes Back",
            ],
          },
        ],
        "searchPeopleByName": [
          "[Person]",
          {
            "search": 5,
          },
        ],
        "searchPlanetsByName": [
          "[Planet]",
          {
            "search": 5,
          },
        ],
        "searchSpeciesByName": [
          "[Species]",
          {
            "search": 5,
          },
        ],
        "searchStarshipsByName": [
          "[Starship]",
          {
            "search": 5,
          },
        ],
        "searchTransportsByName": [
          "[Transport]",
          {
            "search": 5,
          },
        ],
        "searchVehiclesByName": [
          "[Vehicle]",
          {
            "search": 5,
          },
        ],
        "starship": [
          "Starship",
          {
            "id": 7,
          },
        ],
        "transport": [
          "Transport",
          {
            "id": 7,
          },
        ],
        "vehicle": [
          "Vehicle",
          {
            "id": 7,
          },
        ],
      },
      [],
    ],
    "SearchResult": [
      4,
      [
        "Person",
        "Starship",
        "Transport",
        "Species",
        "Vehicle",
        "Planet",
        "Film",
      ],
    ],
    "Species": [
      2,
      {
        "average_height": 2,
        "average_lifespan": 2,
        "classification": 0,
        "created": 0,
        "designation": 0,
        "edited": 0,
        "eye_colors": 0,
        "hair_colors": 0,
        "homeworld": "Planet",
        "id": 2,
        "language": 0,
        "name": 0,
        "people": "[Person]",
        "skin_colors": 0,
      },
      [
        "Node",
      ],
    ],
    "Starship": [
      2,
      {
        "MGLT": 2,
        "cargo_capacity": 2,
        "consumables": 0,
        "cost_in_credits": 2,
        "created": 0,
        "crew": 0,
        "edited": 0,
        "films": "[Film]",
        "hyperdrive_rating": 3,
        "id": 2,
        "image": 0,
        "length": 2,
        "manufacturer": 0,
        "max_atmosphering_speed": 2,
        "model": 0,
        "name": 0,
        "passengers": 2,
        "pilots": "[Person]",
        "starship_class": 0,
      },
      [
        "Node",
      ],
    ],
    "Subscription": [
      2,
      {
        "emitPersons": [
          "Person",
          {
            "limit": 7,
            "throwError": 1,
          },
        ],
      },
      [],
    ],
    "Transport": [
      2,
      {
        "cargo_capacity": 2,
        "consumables": 0,
        "cost_in_credits": 2,
        "created": 0,
        "crew": 0,
        "edited": 0,
        "id": 2,
        "image": 0,
        "length": 2,
        "manufacturer": 0,
        "max_atmosphering_speed": 2,
        "model": 0,
        "name": 0,
        "passengers": 2,
      },
      [
        "Node",
      ],
    ],
    "Vehicle": [
      2,
      {
        "cargo_capacity": 2,
        "consumables": 0,
        "cost_in_credits": 2,
        "created": 0,
        "crew": 2,
        "edited": 0,
        "id": 2,
        "image": 0,
        "length": 3,
        "manufacturer": 0,
        "max_atmosphering_speed": 2,
        "model": 0,
        "name": 0,
        "passengers": 2,
        "pilots": "[Person]",
        "vehicle_class": 0,
      },
      [
        "Node",
      ],
    ],
  },
}
`;
